README.md
Dynamic DynamoDB
================

<a href="http://dynamic-dynamodb.readthedocs.org/en/latest/"><img src="https://readthedocs.org/projects/dynamic-dynamodb/badge/?version=latest"></a>

AWS NoSQL database DynamoDB is a great service, but it lacks automated throughput scaling. This is where Dynamic DynamoDB enters the stage. It provides automatic read and write provisioning for DynamoDB.

All you need to do is to tell Dynamic DynamoDB is at which point and how much you want to scale up or down your DynamoDB tables. An example is in place. Let’s say you have way more traffic on your database during sales hours 4pm - 10pm. DynamicDB can monitor the increased throughput on your DynamoDB instance (via CloudWatch) and provision more throughput as needed. When the load is reducing Dynamic DynamoDB will sense that and automatically reduce your provisioning.

See an example of how to configure Dynamic DynamoDB under **Basic usage** or checkout `dynamic-dynamodb --help`.

Features in short
-----------------

- Scale up and down DynamoDB automatically
- Restrict scaling to certain time slots
- Monitor multiple DynamoDB tables at the same time
- Gives you control over how much reads and writes you want to scale up and down with
- Dynamic DynamoDB has support for max and min limits so that you always knows how much money you spend at most and how much capacity you can be guaranteed
- Support for circuit breaker API call. If your service is experiencing disturbances, Dynamic DynamoDB will not scale down your DynamoDB tables

Documentation
-------------

Project documentation is hosted at [dynamic-dynamodb.readthedocs.org](http://dynamic-dynamodb.readthedocs.org/en/latest/index.html).

Basic usage
-----------

This example will configure Dynamic DynamoDB to:

- Scale up your DynamoDB table when the consumed reads 90% of the total provisioned reads
- Scale up your DynamoDB table when the consumed writes 90% of the total provisioned writes
- Scale up your reads with 50%
- Scale up your writes with 40%
- Scale down your DynamoDB table when the consumed reads 30% of the total provisioned reads
- Scale down your DynamoDB table when the consumed writes 40% of the total provisioned writes
- Scale down your reads with 40%
- Scale down your writes with 70%
- Check for changes every 5 minutes

Command:

    dynamic-dynamodb --table-name my-table \
                     --reads-upper-threshold 90 \
                     --reads-lower-threshold 30 \
                     --increase-reads-with 50 \
                     --decrease-reads-with 40 \
                     --writes-upper-threshold 90 \
                     --writes-lower-threshold 40 \
                     --increase-writes-with 40 \
                     --decrease-writes-with 70 \
                     --check-interval 300

Please note that using configuration files instead of command line options will give you even more control over the service.

Installation instructions
-------------------------

The easiest way to install Dynamic DynamoDB is through PyPI:

    pip install dynamic-dynamodb


Required privileges
-------------------

If you want to set up a separate IAM user for Dynamic DynamoDB, then you need to grant the user the following privileges:

* `cloudwatch:GetMetricStatistics`
* `dynamodb:DescribeTable`
* `dynamodb:ListTables`
* `dynamodb:UpdateTable`
* `sns:Publish` (used by the SNS notifications feature)

An example policy could look like this:

    {
      "Version": "2012-10-17",
      "Statement": [
        {
          "Effect": "Allow",
          "Action": [
            "dynamodb:DescribeTable",
            "dynamodb:ListTables",
            "dynamodb:UpdateTable",
            "cloudwatch:GetMetricStatistics"
          ],
          "Resource": [
            "*"
          ]
        },
        {
          "Effect": "Allow",
          "Action": [
            "sns:Publish"
          ],
          "Resource": [
            "arn:aws:sns:*::dynamic-dynamodb"
          ]
        }
      ]
    }

Community applications
----------------------

There are a few community applications for managing Dynamic DynamoDB.

- [puppet-dynamicdynamodb](https://github.com/mindcandy/puppet-dynamicdynamodb) (Puppet module)
- [elasticDynamoDB](https://github.com/innovia/ElasticDynamoDb) (Wrapper for handling larger planned spikes)
- [Dynamic DynamoDB manager](https://github.com/Mollom/dynamic-dynamodb-manager) (This tool will create your tables, rotate them and update dynamic-dynamodb config file so they are monitored)
- [chef-dynamic-dynamodb](https://github.com/spaceapegames/chef-dynamic-dynamodb) (Chef cookbook - DEPRECATED)

Reporting bugs
--------------

Please help me by providing feedback and bug reports. You can file bugs in the project's [GitHub Issues page](https://github.com/sebdah/dynamic-dynamodb/issues).

Provide as much details as possible to make bug fixing as swift as possible.

Author
------

This project is maintained by [Sebastian Dahlgren](http://www.sebastiandahlgren.se) ([GitHub](https://github.com/sebdah) | [Twitter](https://twitter.com/sebdah) | [LinkedIn](http://www.linkedin.com/in/sebastiandahlgren))

License
-------

APACHE LICENSE 2.0
Copyright 2013-2014 Sebastian Dahlgren

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


docs/release_notes.rst
Release notes
=============

2.5.1
-----

**Release date** 2017-08-12

- Fix `logutils` version to 0.3.3 to prevent pip bug (`#324 <https://github.com/sebdah/dynamic-dynamodb/issues/324>`__).

2.5.0
-----

**Release date** 2017-08-10

- Prevents scaling down write capacity as long as there are throttled write requests (`#319 <https://github.com/sebdah/dynamic-dynamodb/pull/319>`__).
- Print GSI changes as integers instead of floats (`#323 <https://github.com/sebdah/dynamic-dynamodb/pull/323>`__).
- Log table name together with the GSI name for uniqueness (`#331 <https://github.com/sebdah/dynamic-dynamodb/pull/331>`__).
- Documentation fixes (`#329 <https://github.com/sebdah/dynamic-dynamodb/pull/329>`__)

2.4.0
-----

**Release date** 2016-11-25

- Add `--show-config` option (`#317 <https://github.com/sebdah/dynamic-dynamodb/pull/317>`__). Thanks for the pull request `michaelkaye <https://github.com/michaelkaye>`__

2.3.1
-----

**Release date** 2016-08-03

- Bug fix: Include `HTTP 200` as an acceptable response code for the circuit breaker

2.3.0
-----

**Release date** 2016-07-16

- Accept any HTTP `2xx` code as response for a healthy circuit breaker (`#292 <https://github.com/sebdah/dynamic-dynamodb/issues/292>`__)
- Allow `]` in configuration section names (`#281 <https://github.com/sebdah/dynamic-dynamodb/issues/281>`__)
- Security fix for pid and log files (`#301 <https://github.com/sebdah/dynamic-dynamodb/pull/301>`__)
- Update CloudFormation template to allow dots in key-pair name (`#287 <https://github.com/sebdah/dynamic-dynamodb/pull/287>`__)
- Update CloudFormation template to add support for China (Beijing) region (`#288 <https://github.com/sebdah/dynamic-dynamodb/pull/288>`__)
- Fixed `allow-scaling-down-on-0-percent` bug (`#304 <https://github.com/sebdah/dynamic-dynamodb/pull/304>`__)

2.2.1
-----

**Release date** 2015-12-18

- Make sure we aren't downscaling if disabled in the config (`#279 <https://github.com/sebdah/dynamic-dynamodb/pull/279>`__). Thanks for the pull request `Sazpaimon <https://github.com/Sazpaimon>`__

2.2.0
-----

**Release date:** 2015-10-23

- Circuit breaker a per-table/gsi config option (`#276 <https://github.com/sebdah/dynamic-dynamodb/issues/276>`__). Thanks for the pull request `@Sazpaimon <https://github.com/Sazpaimon>`__
- Fix minor spelling and formatting issues (`#278 <https://github.com/sebdah/dynamic-dynamodb/issues/278>`__)

2.1.2
-----

**Release date:** 2015-10-05

- Fix the lower_threshold checking (`#275 <https://github.com/sebdah/dynamic-dynamodb/issues/275>`__). Thanks for the pull request `@KyleAlanDavis <https://github.com/KyleAlanDavis>`__

2.1.1
-----

**Release date:** 2015-10-02

- Depend on argparse for Python 2.6 (`#274 <https://github.com/sebdah/dynamic-dynamodb/issues/274>`__)
- Fix import of ordered dicts for Python 2.6 (`#273 <https://github.com/sebdah/dynamic-dynamodb/issues/273>`__). Thanks for the pull request `@KyleAlanDavis <https://github.com/KyleAlanDavis>`__
- Fix upscaling reads and writes for GSI (`#271 <https://github.com/sebdah/dynamic-dynamodb/issues/271>`__). Thanks for the pull request `@qqshfox <https://github.com/qqshfox>`__

2.1.0
-----

**Release date:** 2015-09-30

- Support for granular downscaling (`#263 <https://github.com/sebdah/dynamic-dynamodb/issues/263>`__). Thanks for the pull request `@Sazpaimon <https://github.com/Sazpaimon>`__
- Allow configurable lookback period instead of the hardcoded 5 minute one (`#261 <https://github.com/sebdah/dynamic-dynamodb/issues/261>`__). Thanks for the pull request `@Sazpaimon <https://github.com/Sazpaimon>`__
- Don't scale down if it would scale below the current consumed units (`#268 <https://github.com/sebdah/dynamic-dynamodb/issues/261>`__). Thanks for the pull request `@Sazpaimon <https://github.com/Sazpaimon>`__

2.0.2
-----

**Release date:** 2015-09-30

- Use HTTP rather than SSH for git clone (`#266 <https://github.com/sebdah/dynamic-dynamodb/issues/266>`__). Thanks for the pull request `@superprat <https://github.com/superprat>`__

2.0.1
-----

**Release date:** 2015-09-29

- Fix value error when logging floats (`#265 <https://github.com/sebdah/dynamic-dynamodb/issues/265>`__)
- Fix installation on Python < 2.7 (`#260 <https://github.com/sebdah/dynamic-dynamodb/issues/260>`__). Thanks for the pull request `@Sazpaimon <https://github.com/Sazpaimon>`__

2.0.0
-----

**Release date:** 2015-08-26

- Match each table to at most one config section (`#250 <https://github.com/sebdah/dynamic-dynamodb/issues/250>`__). Thanks for the pull request `@memory <https://github.com/memory>`__
- Granular up-scaling and new scaling metrics (`#252 <https://github.com/sebdah/dynamic-dynamodb/issues/250>`). Thanks for the pull request `@omnidavez <https://github.com/omnidavez>`__

Please note that this change is not backwards compatible due to the nature of `#250 <https://github.com/sebdah/dynamic-dynamodb/issues/250>__`. Though it is in a minority of use cases, it may in some cases break existing configuration. Details can be found in the pull request `#251 <https://github.com/sebdah/dynamic-dynamodb/issues/251>`__

1.20.6
------

**Release date:** 2015-06-02

- Logging an AccessDeniedException in dynamodb.update_table_provisioning causing a KeyError (`#247 <https://github.com/sebdah/dynamic-dynamodb/issues/247>`__)

1.20.5
------

**Release date:** 2015-04-25

- Update cloudformation template to work around pip permissions issue (`#238 <https://github.com/sebdah/dynamic-dynamodb/issues/238>`__). Thanks for the pull request `@jasonrdsouza <https://github.com/jasonrdsouza>`__

1.20.4
------

**Release date:** 2015-03-17

- Does not scale down when one of the provisoned values is already at the lowest limit (`#237 <https://github.com/sebdah/dynamic-dynamodb/pull/237>`__). Thanks for the pull request `@lcabral37 <https://github.com/lcabral37>`__

1.20.3
------

**Release date:** 2015-02-17


dynamic_dynamodb_old/main.py
# -*- coding: utf-8 -*-
"""
Dynamic DynamoDB

Auto provisioning functionality for Amazon Web Service DynamoDB tables.


APACHE LICENSE 2.0
Copyright 2013 Sebastian Dahlgren

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""
import sys
import logger
import argparse
import ConfigParser
import dynamic_dynamodb

from . import version
from dynamic_dynamodb_daemon import DynamicDynamoDBDaemon


def main():
    """ Main function handling option parsing etc """
    parser = argparse.ArgumentParser(
        description='Dynamic DynamoDB - Auto provisioning AWS DynamoDB')
    parser.add_argument('-c', '--config',
        help='Read configuration from a configuration file')
    parser.add_argument('--dry-run',
        action='store_true',
        help='Run without making any changes to your DynamoDB table')
    parser.add_argument('--daemon',
        help='Run Dynamic DynamoDB as a daemon [start|stop|restart]')
    parser.add_argument('--check-interval',
        type=int,
        default=300,
        help="""How many seconds should we wait between
                the checks (default: 300)""")
    parser.add_argument('--log-file',
        help='Send output to the given log file')
    parser.add_argument('--version',
        action='store_true',
        help='Print current version number')
    parser.add_argument('--aws-access-key-id',
        default=None,
        help="Override Boto configuration with the following AWS access key")
    parser.add_argument('--aws-secret-access-key',
        default=None,
        help="Override Boto configuration with the following AWS secret key")
    dynamodb_ag = parser.add_argument_group('DynamoDB settings')
    dynamodb_ag.add_argument('-r', '--region',
        default='us-east-1',
        help='AWS region to operate in (default: us-east-1')
    dynamodb_ag.add_argument('-t', '--table-name',
        help='How many percent should we decrease the read units with?')
    r_scaling_ag = parser.add_argument_group('Read units scaling properties')
    r_scaling_ag.add_argument('--reads-upper-threshold',
        default=90,
        type=int,
        help="""Scale up the reads with --increase-reads-with percent if
                the currently consumed read units reaches this many
                percent (default: 90)""")
    r_scaling_ag.add_argument('--reads-lower-threshold',
        default=30,
        type=int,
        help="""Scale down the reads with --decrease-reads-with percent if the
                currently consumed read units is as low as this
                percentage (default: 30)""")
    r_scaling_ag.add_argument('--increase-reads-with',
        default=50,
        type=int,
        help="""How many percent should we increase the read
                units with? (default: 50, max: 100)""")
    r_scaling_ag.add_argument('--decrease-reads-with',
        default=50,
        type=int,
        help="""How many percent should we decrease the
                read units with? (default: 50)""")
    r_scaling_ag.add_argument('--min-provisioned-reads',
        type=int,
        help="""Minimum number of provisioned reads""")
    r_scaling_ag.add_argument('--max-provisioned-reads',
        type=int,
        help="""Maximum number of provisioned reads""")
    w_scaling_ag = parser.add_argument_group('Write units scaling properties')
    w_scaling_ag.add_argument('--writes-upper-threshold',
        default=90,
        type=int,
        help="""Scale up the writes with --increase-writes-with percent
                if the currently consumed write units reaches this
                many percent (default: 90)""")
    w_scaling_ag.add_argument('--writes-lower-threshold',
        default=30,
        type=int,
        help="""Scale down the writes with --decrease-writes-with percent
                if the currently consumed write units is as low as this
                percentage (default: 30)""")
    w_scaling_ag.add_argument('--increase-writes-with',
        default=50,
        type=int,
        help="""How many percent should we increase the write
                units with? (default: 50, max: 100)""")
    w_scaling_ag.add_argument('--decrease-writes-with',
        default=50,
        type=int,
        help="""How many percent should we decrease the write
                units with? (default: 50)""")
    w_scaling_ag.add_argument('--min-provisioned-writes',
        type=int,
        help="""Minimum number of provisioned writes""")
    w_scaling_ag.add_argument('--max-provisioned-writes',
        type=int,
        help="""Maximum number of provisioned writes""")
    args = parser.parse_args()

    if args.version:
        print 'Dynamic DynamoDB version: {0}'.format(version)
        sys.exit(0)

    if args.aws_access_key_id and not args.aws_secret_access_key:
        print ('Both --aws-access-key-id and --aws-secret-access-key must '
            'be specified.')
        parser.print_help()
        sys.exit(1)

    if args.config:
        # Handle configuration file
        config = parse_configuration_file(args.config)
        log_level = get_config_log_level(args.config)
        log_file = get_config_log_file(args.config)
        log_handler = logger.Logger(
            level=log_level,
            log_file=log_file,
            dry_run=args.dry_run)
    else:
        # Handle command line arguments
        if not args.table_name:
            print 'argument -t/--table-name is required'
            parser.print_help()
            sys.exit(1)

        config = {}
        config['region'] = args.region
        config['table-name'] = args.table_name
        config['reads-upper-threshold'] = args.reads_upper_threshold
        config['reads-lower-threshold'] = args.reads_lower_threshold
        config['increase-reads-with'] = args.increase_reads_with
        config['decrease-reads-with'] = args.decrease_reads_with
        config['writes-upper-threshold'] = args.writes_upper_threshold
        config['writes-lower-threshold'] = args.writes_lower_threshold
        config['increase-writes-with'] = args.increase_writes_with
        config['decrease-writes-with'] = args.decrease_writes_with
        config['min-provisioned-reads'] = args.min_provisioned_reads
        config['max-provisioned-reads'] = args.max_provisioned_reads
        config['min-provisioned-writes'] = args.min_provisioned_writes
        config['max-provisioned-writes'] = args.max_provisioned_writes
        config['allow-scaling-down-reads-on-0-percent'] = False
        config['allow-scaling-down-writes-on-0-percent'] = False
        config['check-interval'] = args.check_interval
        config['aws-access-key-id'] = args.aws_access_key_id
        config['aws-secret-access-key'] = args.aws_secret_access_key
        config['maintenance-windows'] = None
        if args.log_file:
            log_handler = logger.Logger(
                dry_run=args.dry_run,
                log_file=args.log_file)
        else:
            log_handler = logger.Logger(dry_run=args.dry_run)

    # Options that can only be seet on command line:
    config['dry-run'] = args.dry_run

    if args.daemon:
        daemon = DynamicDynamoDBDaemon('/tmp/daemon.pid')

        if args.daemon == 'start':
            daemon.start(
                config['region'],
                config['table-name'],
                config['reads-upper-threshold'],
                config['reads-lower-threshold'],
                config['increase-reads-with'],
                config['decrease-reads-with'],
                config['writes-upper-threshold'],
                config['writes-lower-threshold'],
                config['increase-writes-with'],
                config['decrease-writes-with'],
                config['min-provisioned-reads'],
                config['max-provisioned-reads'],
                config['min-provisioned-writes'],
                config['max-provisioned-writes'],
                config['allow-scaling-down-reads-on-0-percent'],
                config['allow-scaling-down-writes-on-0-percent'],
                config['always-decrease-rw-together'],
                check_interval=config['check-interval'],
                dry_run=config['dry-run'],
                aws_access_key_id=config['aws-access-key-id'],
                aws_secret_access_key=config['aws-secret-access-key'],
                maintenance_windows=config['maintenance-windows'],
                logger=log_handler)
        elif args.daemon == 'stop':
            daemon.stop()
        elif args.daemon == 'restart':
            daemon.restart()
        else:
            print ('Valid options for --daemon are start, stop and restart')
            parser.print_help()
            sys.exit(1)
    else:
        dynamic_ddb = dynamic_dynamodb.DynamicDynamoDB(
            config['region'],
            config['table-name'],
            config['reads-upper-threshold'],
            config['reads-lower-threshold'],
            config['increase-reads-with'],
            config['decrease-reads-with'],
            config['writes-upper-threshold'],
            config['writes-lower-threshold'],
            config['increase-writes-with'],
            config['decrease-writes-with'],
            config['min-provisioned-reads'],
            config['max-provisioned-reads'],
            config['min-provisioned-writes'],
            config['max-provisioned-writes'],
            config['allow-scaling-down-reads-on-0-percent'],
            config['allow-scaling-down-writes-on-0-percent'],
            config['always-decrease-rw-together'],
            check_interval=config['check-interval'],
            dry_run=config['dry-run'],
            aws_access_key_id=config['aws-access-key-id'],
            aws_secret_access_key=config['aws-secret-access-key'],
            maintenance_windows=config['maintenance-windows'],
            logger=log_handler)
        dynamic_ddb.run()


def get_config_log_level(config_path):
    """ Return the configured log level

    :type config_path: str
    :param config_path: Path to the configuration file
    """
    # Read the configuration file
    config_file = ConfigParser.SafeConfigParser()
    config_file.optionxform = lambda option: option
    config_file.read(config_path)

    try:
        return config_file.get('logging', 'log-level')
    except ConfigParser.NoOptionError:
        return 'info'
    except ConfigParser.NoSectionError:
        return 'info'
    return 'info'


def get_config_log_file(config_path):
    """ Return the configured log file

    :type config_path: str
    :param config_path: Path to the configuration file
    """
    # Read the configuration file
    config_file = ConfigParser.SafeConfigParser()
    config_file.optionxform = lambda option: option
    config_file.read(config_path)

    try:
        return config_file.get('logging', 'log-file')
    except ConfigParser.NoOptionError:
        return None
    except ConfigParser.NoSectionError:
        return None
    return None


def parse_configuration_file(config_path):
    """ Parse a configuration file

    :type config_path: str
    :param config_path: Path to the configuration file
    """
    # Read the configuration file
    config_file = ConfigParser.SafeConfigParser()
    config_file.optionxform = lambda option: option
    config_file.read(config_path)

    # Config dict
    config = {}

    # Find the first table definition
    section = None
    for current_section in config_file.sections():
        if current_section.rsplit(':', 1)[0] != 'table':
            continue
        section = current_section
        config['table-name'] = current_section.rsplit(':', 1)[1].strip()
        break

    if not section:
        print 'Could not find a [table: <table_name>] section in {0}'.format(
            config_path)
        sys.exit(1)

    # Global options to consider
    global_options = [
        ('region', True),
        ('check-interval', True),
        ('aws-access-key-id', False),
        ('aws-secret-access-key', False),
    ]

    # Populate the global options
    for option, required in global_options:
        try:
            config[option] = config_file.get('global', option)
        except ConfigParser.NoOptionError:
            if required:
                print 'Missing [global] option "{0}" in {1}'.format(
                    option, config_path)
                sys.exit(1)
            else:
                config[option] = None
        except ConfigParser.NoSectionError:
            print 'Missing section [global] in {0}'.format(config_path)
            sys.exit(1)

    # Table options to consider
    table_options = [
        ('reads-upper-threshold', True),
        ('reads-lower-threshold', True),
        ('increase-reads-with', True),
        ('decrease-reads-with', True),
        ('writes-upper-threshold', True),
        ('writes-lower-threshold', True),
        ('increase-writes-with', True),
        ('decrease-writes-with', True),
        ('min-provisioned-reads', False),
        ('max-provisioned-reads', False),
        ('min-provisioned-writes', False),
        ('max-provisioned-writes', False),
        ('maintenance-windows', False),
        ('allow-scaling-down-reads-on-0-percent', False),
        ('allow-scaling-down-writes-on-0-percent', False),
        ('always-decrease-rw-together', False),
    ]

    # Populate the table options
    for option, required in table_options:
        try:
            if option == 'allow-scaling-down-reads-on-0-percent':
                config[option] = config_file.getboolean(section, option)
            elif option == 'allow-scaling-down-writes-on-0-percent':
                config[option] = config_file.getboolean(section, option)
            elif option == 'always-decrease-rw-together':
                config[option] = config_file.getboolean(section, option)
            else:
                config[option] = config_file.get(section, option)
        except ConfigParser.NoOptionError:
            if required:
                print 'Missing [{0}] option "{1}" in {2}'.format(
                    section, option, config_path)
                sys.exit(1)
            else:
                if option == 'allow-scaling-down-reads-on-0-percent':
                    config[option] = False
                elif option == 'allow-scaling-down-writes-on-0-percent':
                    config[option] = False
                elif option == 'always-decrease-rw-together':
                    config[option] = False
                else:
                    config[option] = None

    return config
